### ISSUE TITLE
konsole is slow to start, stats too many icons

### ISSUE THREAD

**Issue by ivan** [NixOS/nixpkgs#67691]
**Describe the bug**

konsole is slow to start because it tries to stat ~120K icon files. For some icons, it tries to stat every icon size in every theme.

**To Reproduce**
1. Run NixOS master with `services.xserver.desktopManager.xfce4-14.enable = true;` 
2. Run konsole and observe 1 second startup; strace to see way too many stats

I did not test plasma to see if it reproduces there.

**Additional context**

### NixOS master

```
$ /run/current-system/sw/bin/konsole --version
konsole 19.04.1

$ strace /run/current-system/sw/bin/konsole 2>&1 | grep /share/icons/ | wc -l
120065
```

Startup time is ~1s

I also built konsole 18.08.1 (with dependencies left unchanged) and it had the same problem, so it is not a recent regression in konsole itself.

### Ubuntu 19.04

```
$ strace konsole 2>&1 | grep /share/icons/ | wc -l
968
```

Startup time is ~250ms

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.68-hardened, NixOS, 19.09.git.2610a7b (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(root): `"nixos-18.09.1922.97e0d53d669"`
 - channels(at): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

cc maintainer @ttuegel

**Comment by matthewbauer** [comment]
This could be from XDG_DATA_DIRS or XCURSOR_PATH values? Maybe try messing with those to see if it changes things. My values on KDE are smaller though:

```
$ strace konsole 2>&1 | grep /share/icons/ | wc -l
1896
```

**Comment by ivan** [comment]
my XCURSOR_PATH looks small and reasonable, but XDG_DATA_DIRS is large:

```
XDG_DATA_DIRS=/nix/store/3snj4802y1h7hi4asqgdgkd5mhnyxjl8-konsole-19.04.1/share:
/nix/store/6zqkwqqrly2ad2v7pivy82s7ams3ykwg-knewstuff-5.58.0/share:
/nix/store/mx2fir1j9ycl2qvbjn58vdzlhhsk6g34-kpty-5.58.0/share:
/nix/store/ash4mz25d8xma4q0422b0rir7kz2fl5s-knotifyconfig-5.58.0/share:
/nix/store/mjvfhv0ffv5fwb1csgzrfas57m808fs1-kdelibs4support-5.58.0/share:
/nix/store/y35v7nyls21sax25qp055v3x0s7fi6p8-kded-5.58.0/share:
/nix/store/d1f0jzjb70g9jpf1f4292r361hqk93ij-kunitconversion-5.58.0/share:
/nix/store/k0rdy8bshzi5y69chi5sgkzr91ipb23n-kparts-5.58.0/share:
/nix/store/668ipix6kdzs3qx3d6kl8klacbn6bnpz-kparts-5.58.0-bin/share:
/nix/store/6iy3n4lks367xcbqjp5vfxjnfk84546f-kinit-5.58.0/share:
/nix/store/7b5g2dpk8s5gsm4xlrsln0sj92p7x3qc-kemoticons-5.58.0-bin/share:
/nix/store/bf10f4j9f4ba37kwgq60fw22cvcb2hgj-kdesignerplugin-5.58.0/share:
/nix/store/j9n8r138qw9a4pxkjbj63shfb23w6l04-kio-5.58.0/share:
/nix/store/bd7fndbdh6dj5c7jy32g27l8ramybnzl-kwallet-5.58.0/share:
/nix/store/xl6qcg1pkfpl4j9lw2g2h04w7hlg3dya-kwallet-5.58.0-bin/share:
/nix/store/6sm4wsm3bk94xbg8gh9k6rfdb9ppphv7-knotifications-5.58.0/share:
/nix/store/5nampsblpflij5lrc61p4ai9kg507g47-knotifications-5.58.0-bin/share:
/nix/store/jwspwwz8080y7x543wsqh4d6w9qz2a10-solid-5.58.0/share:
/nix/store/ny7l1cvx3hdjlp40c3mzx42fq3xx2c36-kjobwidgets-5.58.0/share:
/nix/store/xq5s47p6fiqirza0l499a0bqdqpy643k-kjobwidgets-5.58.0-bin/share:
/nix/store/mzlh1bg8385bknn3ad5cnv1k9cf6nmyd-kio-5.58.0-bin/share:
/nix/store/kjdn5yds6dys24b0gxy8vf6d4p8pmxaz-phonon-qt5-4.10.2/share:
/nix/store/j09ydniwk35haph0cjb76jys4as4mlcc-kbookmarks-5.58.0/share:
/nix/store/qbajl24idfh820kxdqfnnb64z7dv1xhy-kxmlgui-5.58.0/share:
/nix/store/y7xmg8ni6p6ra270r4fn4b4q5w444hnp-ktextwidgets-5.58.0/share:
/nix/store/60q2f18ysf2v35gs278qq7ndnr4syy3p-ktextwidgets-5.58.0-bin/share:
/nix/store/3n3ybqvpz8ip377sfm7c90znv64vpqs1-kcompletion-5.58.0/share:
/nix/store/0cr32rk9rsklr4i59g2xhk3pzy86f9fy-sonnet-5.58.0/share:
/nix/store/dlxvyn38bpdjylk5c661i5kfiymdnk1p-kglobalaccel-5.58.0/share:
/nix/store/acbinf6w0wpwqj7h1vg5pj6hskjc8fzl-kservice-5.58.0/share:
/nix/store/sx9m06z4k9xhx7a9bixn8zfd0sr80xpl-kservice-5.58.0-bin/share:
/nix/store/mf956gj5985nviy1vkc2hh6swnjwjmvh-kdbusaddons-5.58.0/share:
/nix/store/c4xj4i0c0p5i1w1049n9xdfmqwr88pbk-kiconthemes-5.58.0/share:
/nix/store/f7yqxyzib19i3npy5zliv9r2w8qwi3g6-kitemviews-5.58.0/share:
/nix/store/qng41d9bh1rzx8mxxrk7a1yl7mhyg35d-kconfigwidgets-5.58.0/share:
/nix/store/gbz9flv5lwx0db0ajqaswz261pzpjm14-kcoreaddons-5.58.0/share:
/nix/store/j6szdnvkba86nyv4bllz48glvp2kah2s-kcoreaddons-5.58.0-bin/share:
/nix/store/48nwvvvd2ixjkhv77sl1jgap61qs66az-kconfig-5.58.0/share:
/nix/store/cha47w7pq68g6g08m9ka01jix518ylnw-kcodecs-5.58.0/share:
/nix/store/bc3imc75var3g99wpn30brzk0pgdvk3q-kglobalaccel-5.58.0-bin/share:
/nix/store/xgnlxs20pyj0m15jdzqcnjs0dzydmcq2-kwindowsystem-5.58.0/share:
/nix/store/7n6ywg47jxx45yprai36p2sbs6favhj7-kauth-5.58.0/share:
/nix/store/69vcnsqm9wl1qb27493rrvybhjmxhm66-kwidgetsaddons-5.58.0/share:
/nix/store/f9f5c3x08xy92x5zh2fqi5m7gv0309a4-patchelf-0.9/share:
/nix/store/1hsfp26nld0qsn3rj5b08nz28m5nby0f-flex-2.6.4/share:
/nix/store/g2f25j76c1fk3rfgi47pcmr44lmwpprp-bison-3.4.1/share:
/nix/store/f66n8plr4albh1wy7wwna4x91y0hwsbh-kdoctools-5.58.0/share:
/nix/store/j23hkszsnkrh34cvy1j08g0s5hzkdbc2-ki18n-5.58.0/share:
/nix/store/z1fmvm5wmhsy7d52h8bknnzsx1cgq2iw-qtbase-5.12.0/share:
/nix/store/my5a5a07x9p5lhvc2n3ki6l20dfz38zh-systemd-242/share:
/nix/store/qj5mxxzcwksccq69zlq9n49d8j4nvpj4-systemd-242-lib/share:
/nix/store/f9skp1y3jr0658b2vgnnj3ginvj71x9p-glib-2.60.4/share:
/nix/store/3njdywpyqns813nc5f2padax9vsdvyl5-gettext-0.19.8.1/share:
/nix/store/5x82ydyjnbcvdw4fpwx7170qfbf2i4x5-dbus-1.12.16/share:
/nix/store/1nrimml11nlg7spkrq0ki1337cjc4222-expat-2.2.7/share:
/nix/store/ji9kpgmbddkv2l2szzjfqinhja2322k0-xz-5.2.4/share:
/nix/store/hrb88p3rriaqzwcpc7asgfbky0smdsds-kdoctools-5.58.0-dev/share:
/nix/store/vjznmqrvwiawqy7g2ip6jsycyls0zlfz-pkg-config-0.29.2/share:
/nix/store/c8waarcjyhgq8ripcrcnc5xqq792zd9c-cmake-3.14.5/share:
/nix/store/3snj4802y1h7hi4asqgdgkd5mhnyxjl8-konsole-19.04.1/share:
/nix/store/3snj4802y1h7hi4asqgdgkd5mhnyxjl8-konsole-19.04.1/share:
/nix/store/6zqkwqqrly2ad2v7pivy82s7ams3ykwg-knewstuff-5.58.0/share:
/nix/store/mx2fir1j9ycl2qvbjn58vdzlhhsk6g34-kpty-5.58.0/share:
/nix/store/ash4mz25d8xma4q0422b0rir7kz2fl5s-knotifyconfig-5.58.0/share:
/nix/store/mjvfhv0ffv5fwb1csgzrfas57m808fs1-kdelibs4support-5.58.0/share:
/nix/store/y35v7nyls21sax25qp055v3x0s7fi6p8-kded-5.58.0/share:
/nix/store/d1f0jzjb70g9jpf1f4292r361hqk93ij-kunitconversion-5.58.0/share:
/nix/store/k0rdy8bshzi5y69chi5sgkzr91ipb23n-kparts-5.58.0/share:
/nix/store/668ipix6kdzs3qx3d6kl8klacbn6bnpz-kparts-5.58.0-bin/share:
/nix/store/6iy3n4lks367xcbqjp5vfxjnfk84546f-kinit-5.58.0/share:
/nix/store/7b5g2dpk8s5gsm4xlrsln0sj92p7x3qc-kemoticons-5.58.0-bin/share:
/nix/store/bf10f4j9f4ba37kwgq60fw22cvcb2hgj-kdesignerplugin-5.58.0/share:
/nix/store/j9n8r138qw9a4pxkjbj63shfb23w6l04-kio-5.58.0/share:
/nix/store/bd7fndbdh6dj5c7jy32g27l8ramybnzl-kwallet-5.58.0/share:
/nix/store/xl6qcg1pkfpl4j9lw2g2h04w7hlg3dya-kwallet-5.58.0-bin/share:
/nix/store/6sm4wsm3bk94xbg8gh9k6rfdb9ppphv7-knotifications-5.58.0/share:
/nix/store/5nampsblpflij5lrc61p4ai9kg507g47-knotifications-5.58.0-bin/share:
/nix/store/jwspwwz8080y7x543wsqh4d6w9qz2a10-solid-5.58.0/share:
/nix/store/ny7l1cvx3hdjlp40c3mzx42fq3xx2c36-kjobwidgets-5.58.0/share:
/nix/store/xq5s47p6fiqirza0l499a0bqdqpy643k-kjobwidgets-5.58.0-bin/share:
/nix/store/mzlh1bg8385bknn3ad5cnv1k9cf6nmyd-kio-5.58.0-bin/share:
/nix/store/kjdn5yds6dys24b0gxy8vf6d4p8pmxaz-phonon-qt5-4.10.2/share:
/nix/store/j09ydniwk35haph0cjb76jys4as4mlcc-kbookmarks-5.58.0/share:
/nix/store/qbajl24idfh820kxdqfnnb64z7dv1xhy-kxmlgui-5.58.0/share:
/nix/store/y7xmg8ni6p6ra270r4fn4b4q5w444hnp-ktextwidgets-5.58.0/share:
/nix/store/60q2f18ysf2v35gs278qq7ndnr4syy3p-ktextwidgets-5.58.0-bin/share:
/nix/store/3n3ybqvpz8ip377sfm7c90znv64vpqs1-kcompletion-5.58.0/share:
/nix/store/0cr32rk9rsklr4i59g2xhk3pzy86f9fy-sonnet-5.58.0/share:
/nix/store/dlxvyn38bpdjylk5c661i5kfiymdnk1p-kglobalaccel-5.58.0/share:
/nix/store/acbinf6w0wpwqj7h1vg5pj6hskjc8fzl-kservice-5.58.0/share:
/nix/store/sx9m06z4k9xhx7a9bixn8zfd0sr80xpl-kservice-5.58.0-bin/share:
/nix/store/mf956gj5985nviy1vkc2hh6swnjwjmvh-kdbusaddons-5.58.0/share:
/nix/store/c4xj4i0c0p5i1w1049n9xdfmqwr88pbk-kiconthemes-5.58.0/share:
/nix/store/f7yqxyzib19i3npy5zliv9r2w8qwi3g6-kitemviews-5.58.0/share:
/nix/store/qng41d9bh1rzx8mxxrk7a1yl7mhyg35d-kconfigwidgets-5.58.0/share:
/nix/store/gbz9flv5lwx0db0ajqaswz261pzpjm14-kcoreaddons-5.58.0/share:
/nix/store/j6szdnvkba86nyv4bllz48glvp2kah2s-kcoreaddons-5.58.0-bin/share:
/nix/store/48nwvvvd2ixjkhv77sl1jgap61qs66az-kconfig-5.58.0/share:
/nix/store/cha47w7pq68g6g08m9ka01jix518ylnw-kcodecs-5.58.0/share:
/nix/store/bc3imc75var3g99wpn30brzk0pgdvk3q-kglobalaccel-5.58.0-bin/share:
/nix/store/xgnlxs20pyj0m15jdzqcnjs0dzydmcq2-kwindowsystem-5.58.0/share:
/nix/store/7n6ywg47jxx45yprai36p2sbs6favhj7-kauth-5.58.0/share:
/nix/store/69vcnsqm9wl1qb27493rrvybhjmxhm66-kwidgetsaddons-5.58.0/share:
/nix/store/f9f5c3x08xy92x5zh2fqi5m7gv0309a4-patchelf-0.9/share:
/nix/store/1hsfp26nld0qsn3rj5b08nz28m5nby0f-flex-2.6.4/share:
/nix/store/g2f25j76c1fk3rfgi47pcmr44lmwpprp-bison-3.4.1/share:
/nix/store/f66n8plr4albh1wy7wwna4x91y0hwsbh-kdoctools-5.58.0/share:
/nix/store/j23hkszsnkrh34cvy1j08g0s5hzkdbc2-ki18n-5.58.0/share:
/nix/store/z1fmvm5wmhsy7d52h8bknnzsx1cgq2iw-qtbase-5.12.0/share:
/nix/store/my5a5a07x9p5lhvc2n3ki6l20dfz38zh-systemd-242/share:
/nix/store/qj5mxxzcwksccq69zlq9n49d8j4nvpj4-systemd-242-lib/share:
/nix/store/f9skp1y3jr0658b2vgnnj3ginvj71x9p-glib-2.60.4/share:
/nix/store/3njdywpyqns813nc5f2padax9vsdvyl5-gettext-0.19.8.1/share:
/nix/store/5x82ydyjnbcvdw4fpwx7170qfbf2i4x5-dbus-1.12.16/share:
/nix/store/1nrimml11nlg7spkrq0ki1337cjc4222-expat-2.2.7/share:
/nix/store/ji9kpgmbddkv2l2szzjfqinhja2322k0-xz-5.2.4/share:
/nix/store/hrb88p3rriaqzwcpc7asgfbky0smdsds-kdoctools-5.58.0-dev/share:
/nix/store/vjznmqrvwiawqy7g2ip6jsycyls0zlfz-pkg-config-0.29.2/share:
/nix/store/c8waarcjyhgq8ripcrcnc5xqq792zd9c-cmake-3.14.5/share:
/nix/store/3snj4802y1h7hi4asqgdgkd5mhnyxjl8-konsole-19.04.1/share:
/nix/store/4l7yjlsj5dgcqnaynrfz344h5mllilhm-xfce4-session-4.14.0/share:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share:
/nix/store/j1lc5ir64fbrpdhximk2dmi7aykgj92g-hicolor-icon-theme-0.17/share:
/nix/store/ra2ay87kzj07siwvg1lsd4f5azm261ki-exo-0.12.8/share:
/nix/store/adsvkw3891a37nk7xqg2gwfqjlfsb0zr-cups-2.2.11/share:
/nix/store/2vmr74j60jk7dyi53pfihk3ghfd1kz2z-libxfce4ui-4.14.1/share:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share:
/nix/store/j1lc5ir64fbrpdhximk2dmi7aykgj92g-hicolor-icon-theme-0.17/share:
/nix/store/ra2ay87kzj07siwvg1lsd4f5azm261ki-exo-0.12.8/share:
/nix/store/adsvkw3891a37nk7xqg2gwfqjlfsb0zr-cups-2.2.11/share:
/nix/store/2vmr74j60jk7dyi53pfihk3ghfd1kz2z-libxfce4ui-4.14.1/share:
/nix/store/4l7yjlsj5dgcqnaynrfz344h5mllilhm-xfce4-session-4.14.0/share:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/vjayily6sjggjj1qdqmg9713rvpl5wvz-gsettings-desktop-schemas-3.32.0/share/gsettings-schemas/gsettings-desktop-schemas-3.32.0:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share:
/nix/store/j1lc5ir64fbrpdhximk2dmi7aykgj92g-hicolor-icon-theme-0.17/share:
/nix/store/ra2ay87kzj07siwvg1lsd4f5azm261ki-exo-0.12.8/share:
/nix/store/adsvkw3891a37nk7xqg2gwfqjlfsb0zr-cups-2.2.11/share:
/nix/store/2vmr74j60jk7dyi53pfihk3ghfd1kz2z-libxfce4ui-4.14.1/share:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share:
/nix/store/j1lc5ir64fbrpdhximk2dmi7aykgj92g-hicolor-icon-theme-0.17/share:
/nix/store/ra2ay87kzj07siwvg1lsd4f5azm261ki-exo-0.12.8/share:
/nix/store/adsvkw3891a37nk7xqg2gwfqjlfsb0zr-cups-2.2.11/share:
/nix/store/2vmr74j60jk7dyi53pfihk3ghfd1kz2z-libxfce4ui-4.14.1/share:
/home/at/.nix-profile/share:
/etc/profiles/per-user/at/share:
/nix/var/nix/profiles/default/share:
/run/current-system/sw/share:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10:
/nix/store/zhzgdnnaf34imjw6cf057fjdvl8nnfa6-gtk+3-3.24.10/share/gsettings-schemas/gtk+3-3.24.10
```

**Comment by matthewbauer** [comment]
I suspect this is fallout from https://github.com/NixOS/nixpkgs/pull/54525.

**Comment by ttuegel** [comment]
I notice two separate bugs:

1. There are some build dependencies in `XDG_DATA_DIRS`.
1. About half the entries in `XDG_DATA_DIRS` are redundant; there should not be XDG data in both the `$out` and `$bin` output, yet we are searching both.

**Comment by ambrop72** [comment]
I think what we need is the capability in `makeWrapper` and similar to add to a list-type variable only if the value is not already inside.

**Comment by lheckemann** [comment]
I believe this is fixed by #69426/#69427. Feel free to reopen or let me know if I'm mistaken.

**Comment by ivan** [comment]
I cannot reopen, but this is not fixed in xfce4 on release-19.09 (I cannot test master right now, sorry):

```
# strace /run/current-system/sw/bin/konsole 2>&1 | grep /share/icons/ | wc -l
172132
```

**Comment by ivan** [comment]
Oh, I see the problem, I was using konsole to start another konsole, and `XDG_DATA_DIRS` becomes longer each time.
